WooCommerce/Classes/ServiceLocator/ProductImageUploader.swift

```diff
@@ -22,16 +43,57 @@ final class ProductImageUploader: ProductImageUploaderProtocol {
     }
 
     private var actionHandlersByProduct: [ProductKey: ProductImageActionHandler] = [:]
+    private var imagesSaverByProduct: [ProductKey: ProductImagesSaver] = [:]
+    private let stores: StoresManager
+
+    init(stores: StoresManager = ServiceLocator.stores) {
+        self.stores = stores
+    }
 
     func actionHandler(siteID: Int64, productID: Int64, isLocalID: Bool, originalStatuses: [ProductImageStatus]) -> ProductImageActionHandler {
         let key = ProductKey(siteID: siteID, productID: productID, isLocalID: isLocalID)
         let actionHandler: ProductImageActionHandler
         if let handler = actionHandlersByProduct[key], handler.productImageStatuses.hasPendingUpload {
             actionHandler = handler
         } else {
-            actionHandler = ProductImageActionHandler(siteID: siteID, productID: productID, imageStatuses: originalStatuses)
+            actionHandler = ProductImageActionHandler(siteID: siteID, productID: productID, imageStatuses: originalStatuses, stores: stores)
             actionHandlersByProduct[key] = actionHandler
         }
         return actionHandler
     }
+
+    func hasUnsavedChangesOnImages(siteID: Int64, productID: Int64, isLocalID: Bool, originalImages: [ProductImage]) -> Bool {
+        let key = ProductKey(siteID: siteID, productID: productID, isLocalID: isLocalID)
+        guard let handler = actionHandlersByProduct[key] else {
+            return false
+        }
+        // If there are images scheduled to be saved, there are no unsaved changes if the image statuses to save match the latest image statuses.
+        if let productImagesSaver = imagesSaverByProduct[key], productImagesSaver.imageStatusesToSave.isNotEmpty {
+            return handler.productImageStatuses != productImagesSaver.imageStatusesToSave
+        } else {
+            return handler.productImageStatuses.hasPendingUpload ||
+            handler.productImageStatuses.images.map { $0.imageID } != originalImages.map { $0.imageID }
```

Comment:
Nit: I think it will help to add comments for the conditions that we check in `else` block as well. 😄

Reply:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/7048/commits/9dd3efde0d2f70d710601cd8abf7ea90f30a8685